File: xaiop-sdk/python/src/xaiop/engine.py
"""In-memory XAIOP engine (isomorphic core)."""

from __future__ import annotations

import copy
import random
import time
from typing import Any

from .clone import clone_json
from .compat import CompatPolicy
from .encode import encode_sync
from .fragment import XaiopFragment
from .merge import (
    format_inject_result,
    merge_json,
    merge_to_json,
    merge_to_xaiop,
    to_mergeable_json,
)
from .parse import parse_sync
from .types import TypeChecker, TypeRegistry, encode_type_schema_frame

PROTOCOL_VERSION = "0.6.0"
SDK_VERSION = "0.15.1"


class XaiopEngine:
    def __init__(self, *, compatibility_mode: bool = False) -> None:
        self._store: dict[str, Any] = {}
        self._seq = 0
        self._compatibility_mode = bool(compatibility_mode)
        self._type_check = False
        self._type_registry = TypeRegistry()
        self._type_violation_hook = None
        self._compat = CompatPolicy()

    @property
    def compatibility_mode(self) -> bool:
        return self._compatibility_mode

    @property
    def type_check(self) -> bool:
        return self._type_check

    def set_compatibility_mode(self, enabled: bool) -> XaiopEngine:
        self._compatibility_mode = bool(enabled)
        if self._compatibility_mode:
            self._type_check = False
        return self

    def set_type_check(self, enabled: bool) -> bool:
        if enabled:
            if self._compatibility_mode:
                return False
            self._type_check = True
            return True
        self._type_check = False
        return True

    @property
    def type_registry(self) -> TypeRegistry:
        return self._type_registry

    def register_type(
        self,
        path: str,
        type_: Any,
        options: dict[str, Any] | None = None,
    ) -> bool:
        return self._type_registry.register(path, type_, options)

    def register_types(
        self,
        mapping: dict[str, Any] | list[tuple[str, Any]],
        options: dict[str, Any] | None = None,
    ) -> None:
        self._type_registry.register_many(mapping, options)

    def register_type_deny(self, path: str, type_: Any) -> bool:
        return self._type_registry.register(path, type_, {"polarity": "deny"})

    def get_registered_type(self, path: str) -> Any:
        return self._type_registry.get(path)

    def export_type_schema(self) -> dict[str, Any]:
        return self._type_registry.snapshot()

    def on_type_violation(self, fn: Any) -> XaiopEngine:
        self._type_violation_hook = fn if callable(fn) else None
        return self

    def encode_type_schema_frame(self) -> str:
        return encode_type_schema_frame(self._type_registry.snapshot())

    def _run_type_check(self, value: Any) -> None:
        if not self._type_check:
            return
        if self._type_registry.size == 0:
            return
        checker = TypeChecker(
            self._type_registry,
            on_violation=self._type_violation_hook,
        )
        checker.check_tree(value)

    def _set_compat_fix(self, fix_id: str, enabled: bool) -> bool:
        if not self._compatibility_mode:
            return False
        return self._compat.set(fix_id, enabled)

    def _parse_compat_arg(self) -> bool | dict[str, bool]:
        return self._compat.snapshot() if self._compatibility_mode else False

    @property
    def compat_forced_root(self) -> bool:
        return self._compat.forcedRoot

    def set_compat_forced_root(self, enabled: bool) -> bool:
        return self._set_compat_fix("forcedRoot", enabled)

    @property
    def compat_rewrite_bare_name_array(self) -> bool:
        return self._compat.rewriteBareNameArray

    def set_compat_rewrite_bare_name_array(self, enabled: bool) -> bool:
        return self._set_compat_fix("rewriteBareNameArray", enabled)

    @property
    def compat_rewrite_enter_line(self) -> bool:
        return self._compat.rewriteEnterLine

    def set_compat_rewrite_enter_line(self, enabled: bool) -> bool:
        return self._set_compat_fix("rewriteEnterLine", enabled)

    @property
    def compat_ignore_bare_leave_at_root(self) -> bool:
        return self._compat.ignoreBareLeaveAtRoot

    def set_compat_ignore_bare_leave_at_root(self, enabled: bool) -> bool:
        return self._set_compat_fix("ignoreBareLeaveAtRoot", enabled)

    @property
    def compat_pop_and_retry(self) -> bool:
        return self._compat.popAndRetry

    def set_compat_pop_and_retry(self, enabled: bool) -> bool:
        return self._set_compat_fix("popAndRetry", enabled)

    @property
    def compat_locate_path_trim(self) -> bool:
        return self._compat.locatePathTrim

    def set_compat_locate_path_trim(self, enabled: bool) -> bool:
        return self._set_compat_fix("locatePathTrim", enabled)

    @property
    def compat_locate_path_strip_spaces(self) -> bool:
        return self._compat.locatePathStripSpaces

    def set_compat_locate_path_strip_spaces(self, enabled: bool) -> bool:
        return self._set_compat_fix("locatePathStripSpaces", enabled)

    @property
    def compat_locate_path_array_suffix(self) -> bool:
        return self._compat.locatePathArraySuffix

    def set_compat_locate_path_array_suffix(self, enabled: bool) -> bool:
        return self._set_compat_fix("locatePathArraySuffix", enabled)

    def upload_sync(self, source: str) -> str:
        value = parse_sync(source, self._parse_compat_arg())
        self._run_type_check(value)
        data_id = _next_id(self._seq + 1)
        self._seq += 1
        self._store[data_id] = value
        return data_id

    def upload_json_sync(self, value: Any, encode_options: dict[str, Any] | None = None) -> str:
        source = encode_sync(value, **(encode_options or {}))
        return self.upload_sync(source)

    def encode_sync(self, value: Any, options: dict[str, Any] | None = None) -> str:
        return encode_sync(value, **(options or {}))

    def merge_to_json_sync(
        self,
        base_json: Any,
        xaiop_source: str,
        options: dict[str, Any] | None = None,
    ) -> Any:
        options = dict(options or {})
        if "compat" not in options:
            options["compat"] = self._parse_compat_arg()
        return merge_to_json(base_json, xaiop_source, options)

    def merge_to_xaiop_sync(
        self,
        base_json: Any,
        xaiop_source: str,
        options: dict[str, Any] | None = None,
    ) -> str:
        options = dict(options or {})
        if "compat" not in options:
            options["compat"] = self._parse_compat_arg()
        return merge_to_xaiop(base_json, xaiop_source, options)

    def inject_xaiop_sync(
        self,
        data_id: str,
        xaiop_source: str,
        options: dict[str, Any] | None = None,
    ) -> Any:
        options = dict(options or {})
        base = self._require_stored(data_id)
        merge_opts = {
            "conflict": options.get("conflict"),
            "compat": options.get("compat", self._parse_compat_arg()),
        }
        merged = merge_to_json(base, xaiop_source, merge_opts)
        self._run_type_check(merged)
        self._store[data_id] = merged
        return format_inject_result(merged, options)

    def inject_json_sync(
        self,
        data_id: str,
        json_value: Any,
        options: dict[str, Any] | None = None,
    ) -> Any:
        options = options or {}
        base = self._require_stored(data_id)
        merged = merge_json(base, json_value, options.get("conflict"))
        self._run_type_check(merged)
        self._store[data_id] = merged
        return format_inject_result(merged, options)

    def _require_stored(self, data_id: str) -> Any:
        if not isinstance(data_id, str) or not data_id:
            raise TypeError("dataId must be a non-empty string")
        if data_id not in self._store:
            raise ValueError(f"unknown data id: {data_id}")
        return to_mergeable_json(self._store[data_id])

    def get_sync(self, data_id: str) -> Any:
        if not isinstance(data_id, str) or not data_id:
            raise TypeError("dataId must be a non-empty string")
        if data_id not in self._store:
            raise ValueError(f"unknown data id: {data_id}")
        value = self._store[data_id]
        if isinstance(value, XaiopFragment):
            return XaiopFragment(copy.deepcopy(value.entries))
        return copy.deepcopy(value)

    @staticmethod
    def parse_sync(source: str, compatibility_mode: bool = False) -> Any:
        return parse_sync(source, compatibility_mode)

    @staticmethod
    def encode_sync_static(value: Any, options: dict[str, Any] | None = None) -> str:
        return encode_sync(value, **(options or {}))

    @staticmethod
    def merge_to_json_static(
        base_json: Any,
        xaiop_source: str,
        options: dict[str, Any] | None = None,
    ) -> Any:
        return merge_to_json(base_json, xaiop_source, options)

    @staticmethod
    def merge_to_xaiop_static(
        base_json: Any,
        xaiop_source: str,
        options: dict[str, Any] | None = None,
    ) -> str:
        return merge_to_xaiop(base_json, xaiop_source, options)

    def has(self, data_id: str) -> bool:
        return data_id in self._store

    def delete(self, data_id: str) -> bool:
        return self._store.pop(data_id, None) is not None

    def clear(self) -> None:
        self._store.clear()


def _next_id(seq: int) -> str:
    return (
        f"xaiop_{seq}_{int(time.time() * 1000):x}_"
        f"{random.randint(0, 0xFFFFFF):06x}"
    )

File: xaiop-sdk/nodejs/src/core/engine.ts
// @ts-nocheck
/**
 * In-memory XAIOP engine (isomorphic core).
 */
import { CompatPolicy } from "./compat.js";
import {
  encodeAsync,
  encodeSync as encodeValueSync,
} from "./encode.js";
import {
  formatInjectResult,
  mergeJson,
  mergeToJson as mergeToJsonValue,
  mergeToXaiop as mergeToXaiopValue,
  toMergeableJson,
} from "./merge.js";
import {
  parseAsync,
  parseSync,
  XaiopFragment,
} from "./parse.js";
import {
  TypeChecker,
  TypeRegistry,
  encodeTypeSchemaFrame,
} from "./types.js";

export const PROTOCOL_VERSION = "0.7.0";
export const SDK_VERSION = "0.16.0";

export class XaiopEngine {
  /**
   * @param {{ compatibilityMode?: boolean }} [options]
   */
  constructor(options = {}) {
    /** @type {Map<string, unknown>} */
    this._store = new Map();
    this._seq = 0;
    /** @type {boolean} */
    this._compatibilityMode = !!options.compatibilityMode;
    /**
     * Type-check flag (default off). May be on only in strict mode
     * (`compatibilityMode === false`). When on, registered types are enforced
     * on upload/inject via TypeChecker (+ optional violation hook).
     * @type {boolean}
     */
    this._typeCheck = false;
    /** @type {TypeRegistry} */
    this._typeRegistry = new TypeRegistry();
    /**
     * @type {((err: import("./types.js").XaiopTypeError, ctx: object) => void)|null}
     */
    this._typeViolationHook = null;
    /** Per-fix flags; defaults all `true`. Take effect only while compatibility mode is on. */
    this._compat = new CompatPolicy();
  }

  /** Whether compatibility mode is enabled for this engine instance. */
  get compatibilityMode() {
    return this._compatibilityMode;
  }

  /**
   * Whether the type-check flag is on. Default `false`.
   * Flag only — does not imply type checking is implemented or running.
   */
  get typeCheck() {
    return this._typeCheck;
  }

  /**
   * Enable or disable compatibility mode.
   * Does **not** reset per-fix flags (they stay at defaults — all on —
   * until toggled individually).
   * Turning compatibility mode **on** clears `typeCheck` (strict-only flag).
   * @param {boolean} enabled
   * @returns {this}
   */
  setCompatibilityMode(enabled) {
    this._compatibilityMode = !!enabled;
    if (this._compatibilityMode) this._typeCheck = false;
    return this;
  }

  /**
   * Enable or disable the type-check flag.
   * May be enabled only in strict mode (`compatibilityMode === false`).
   * When enabled, `upload*` / `inject*` run the type registry checker.
   * @param {boolean} enabled
   * @returns {boolean} whether the requested state was applied
   */
  setTypeCheck(enabled) {
    if (enabled) {
      if (this._compatibilityMode) return false;
      this._typeCheck = true;
      return true;
    }
    this._typeCheck = false;
    return true;
  }

  /** Path → type registry (immutable entries once registered). */
  get typeRegistry() {
    return this._typeRegistry;
  }

  /**
   * Register an allow (whitelist) or deny (blacklist) type for a JSON path.
   * Path style: `data.fork`, `items[0]`. Once registered, the path cannot be changed.
   * @param {string} path
   * @param {unknown} type — `TYPE.*`, `objectType`/`arrayType`, or surface string
   * @param {{ polarity?: 'allow'|'deny' }} [options]
   * @returns {boolean} false if path already registered
   */
  registerType(path, type, options = {}) {
    return this._typeRegistry.register(path, type, options);
  }

  /**
   * Register multiple path/type pairs (same polarity option for object-map form).
   * @param {Record<string, unknown>|Iterable<[string, unknown]>} map
   * @param {{ polarity?: 'allow'|'deny' }} [options]
   */
  registerTypes(map, options = {}) {
    return this._typeRegistry.registerMany(map, options);
  }

  /**
   * Blacklist helper — `registerType(path, type, { polarity: 'deny' })`.
   * @param {string} path
   * @param {unknown} type
   * @returns {boolean}
   */
  registerTypeDeny(path, type) {
    return this._typeRegistry.register(path, type, { polarity: "deny" });
  }

  /** @param {string} path */
  getRegisteredType(path) {
    return this._typeRegistry.get(path);
  }

  /** Export schema snapshot for WS `pushTypeConsistency`. */
  exportTypeSchema() {
    return this._typeRegistry.snapshot();
  }

  /**
   * Hook invoked on each registry violation while typeCheck is on (before throw).
   * @param {((err: import("./types.js").XaiopTypeError, ctx: object) => void)|null} fn
   * @returns {this}
   */
  onTypeViolation(fn) {
    this._typeViolationHook = typeof fn === "function" ? fn : null;
    return this;
  }

  /**
   * Encode a type-schema control frame (for tests / manual WS send).
   * Prefer `XaiopWsConnection.pushTypeConsistency(engine)`.
   */
  encodeTypeSchemaFrame() {
    return encodeTypeSchemaFrame(this._typeRegistry.snapshot());
  }

  /** @param {unknown} value */
  _runTypeCheck(value) {
    if (!this._typeCheck) return;
    if (this._typeRegistry.size === 0) return;
    const checker = new TypeChecker(this._typeRegistry, {
      onViolation: this._typeViolationHook || undefined,
    });
    checker.checkTree(value);
  }

  /**
   * @param {import("./compat.js").CompatFixId} id
   * @param {boolean} enabled
   * @returns {boolean}
   */
  _setCompatFix(id, enabled) {
    if (!this._compatibilityMode) return false;
    return this._compat.set(id, enabled);
  }

  /** @returns {false|ReturnType<CompatPolicy["snapshot"]>} */
  _parseCompatArg() {
    return this._compatibilityMode ? this._compat.snapshot() : false;
  }

  // --- Compatibility fix APIs (active only while compatibilityMode is on) ---

  /** Force a complete anonymous object root when the stream does not open with `>` / `-`. */
  get compatForcedRoot() {
    return this._compat.forcedRoot;
  }
  /** @param {boolean} enabled @returns {boolean} */
  setCompatForcedRoot(enabled) {
    return this._setCompatFix("forcedRoot", enabled);
  }

  /** Rewrite bare `name-` → `>name-`. */
  get compatRewriteBareNameArray() {
    return this._compat.rewriteBareNameArray;
  }
  /** @param {boolean} enabled @returns {boolean} */
  setCompatRewriteBareNameArray(enabled) {
    return this._setCompatFix("rewriteBareNameArray", enabled);
  }

  /**
   * Rewrite `>` whitespace / glued `>key:value` / trailing spaces on Structure lines.
   */
  get compatRewriteEnterLine() {
    return this._compat.rewriteEnterLine;
  }
  /** @param {boolean} enabled @returns {boolean} */
  setCompatRewriteEnterLine(enabled) {
    return this._setCompatFix("rewriteEnterLine", enabled);
  }

  /** Ignore redundant bare `<` while Cursor is already at document Root. */
  get compatIgnoreBareLeaveAtRoot() {
    return this._compat.ignoreBareLeaveAtRoot;
  }
  /** @param {boolean} enabled @returns {boolean} */
  setCompatIgnoreBareLeaveAtRoot(enabled) {
    return this._setCompatFix("ignoreBareLeaveAtRoot", enabled);
  }

  /** On `XaiopSyntaxError`, pop Cursor and retry the same line. */
  get compatPopAndRetry() {
    return this._compat.popAndRetry;
  }
  /** @param {boolean} enabled @returns {boolean} */
  setCompatPopAndRetry(enabled) {
    return this._setCompatFix("popAndRetry", enabled);
  }

  /** On `=path not found`, retry after trimming edge whitespace. */
  get compatLocatePathTrim() {
    return this._compat.locatePathTrim;
  }
  /** @param {boolean} enabled @returns {boolean} */
  setCompatLocatePathTrim(enabled) {
    return this._setCompatFix("locatePathTrim", enabled);
  }

  /** On `=path not found`, retry after stripping all whitespace. */
  get compatLocatePathStripSpaces() {
    return this._compat.locatePathStripSpaces;
  }
  /** @param {boolean} enabled @returns {boolean} */
  setCompatLocatePathStripSpaces(enabled) {
    return this._setCompatFix("locatePathStripSpaces", enabled);
  }

  /**
   * On `=path not found`, treat a segment trailing `-` as `>name-` create postfix
   * (match array keys only).
   */
  get compatLocatePathArraySuffix() {
    return this._compat.locatePathArraySuffix;
  }
  /** @param {boolean} enabled @returns {boolean} */
  setCompatLocatePathArraySuffix(enabled) {
    return this._setCompatFix("locatePathArraySuffix", enabled);
  }

  // --- Store APIs ---

  /**
   * Upload XAIOP text. Returns a runtime data id.
   * @param {string} source
   * @returns {Promise<string>}
   */
  async upload(source) {
    return this.uploadSync(source);
  }

  /**
   * @param {string} source
   * @returns {string}
   */
  uploadSync(source) {
    const value = parseSync(source, this._parseCompatArg());
    this._runTypeCheck(value);
    const id = nextId(++this._seq);
    this._store.set(id, value);
    return id;
  }

  /**
   * Encode JSON → XAIOP, then upload. Returns a runtime data id.
   * Compatibility mode does **not** affect encode (strict wire only).
   * @param {unknown} value
   * @param {import("./encode.js").EncodeOptions} [encodeOptions]
   * @returns {Promise<string>}
   */
  async uploadJson(value, encodeOptions = {}) {
    return this.uploadJsonSync(value, encodeOptions);
  }

  /**
   * @param {unknown} value
   * @param {import("./encode.js").EncodeOptions} [encodeOptions]
   * @returns {string}
   */
  uploadJsonSync(value, encodeOptions = {}) {
    const source = encodeValueSync(value, encodeOptions);
    return this.uploadSync(source);
  }

  /**
   * Instance encode (strict wire). Same options as static `encodeSync`.
   * @param {unknown} value
   * @param {import("./encode.js").EncodeOptions} [options]
   * @returns {Promise<string>}
   */
  async encode(value, options = {}) {
    return encodeValueSync(value, options);
  }

  /** Identical to {@link encode}; named to match free `encodeAsync`. */
  async encodeAsync(value, options = {}) {
    return this.encode(value, options);
  }

  /**
   * @param {unknown} value
   * @param {import("./encode.js").EncodeOptions} [options]
   * @returns {string}
   */
  encodeSync(value, options = {}) {
    return encodeValueSync(value, options);
  }

  // --- Merge / inject (pre/post — not streaming) ---

  /**
   * Merge base JSON + XAIOP → JSON. Uses this engine's compat for parse.
   * @param {unknown} baseJson
   * @param {string} xaiopSource
   * @param {import("./merge.js").MergeOptions} [options]
   */
  async mergeToJson(baseJson, xaiopSource, options = {}) {
    return this.mergeToJsonSync(baseJson, xaiopSource, options);
  }

  /**
   * @param {unknown} baseJson
   * @param {string} xaiopSource
   * @param {import("./merge.js").MergeOptions} [options]
   */
  mergeToJsonSync(baseJson, xaiopSource, options = {}) {
    return mergeToJsonValue(baseJson, xaiopSource, {
      ...options,
      compat: options.compat !== undefined ? options.compat : this._parseCompatArg(),
    });
  }

  /**
   * Merge base JSON + XAIOP → XAIOP wire.
   * @param {unknown} baseJson
   * @param {string} xaiopSource
   * @param {import("./merge.js").MergeToXaiopOptions} [options]
   */
  async mergeToXaiop(baseJson, xaiopSource, options = {}) {
    return this.mergeToXaiopSync(baseJson, xaiopSource, options);
  }

  /**
   * @param {unknown} baseJson
   * @param {string} xaiopSource
   * @param {import("./merge.js").MergeToXaiopOptions} [options]
   */
  mergeToXaiopSync(baseJson, xaiopSource, options = {}) {
    return mergeToXaiopValue(baseJson, xaiopSource, {
      ...options,
      compat: options.compat !== undefined ? options.compat : this._parseCompatArg(),
    });
  }

  /**
   * Inject XAIOP into a stored document (mutates store).
   * @param {string} dataId
   * @param {string} xaiopSource
   * @param {import("./merge.js").InjectOptions} [options]
   * @returns {Promise<unknown|string>}
   */
  async injectXaiop(dataId, xaiopSource, options = {}) {
    return this.injectXaiopSync(dataId, xaiopSource, options);
  }

  /**
   * @param {string} dataId
   * @param {string} xaiopSource
   * @param {import("./merge.js").InjectOptions} [options]
   * @returns {unknown|string}
   */
  injectXaiopSync(dataId, xaiopSource, options = {}) {
    const base = this._requireStored(dataId);
    const merged = mergeToJsonValue(base, xaiopSource, {
      conflict: options.conflict,
      compat:
        options.compat !== undefined ? options.compat : this._parseCompatArg(),
    });
    this._runTypeCheck(merged);
    this._store.set(dataId, merged);
    return formatInjectResult(merged, options);
  }

  /**
   * Inject JSON into a stored document (mutates store).
   * @param {string} dataId
   * @param {unknown} jsonValue
   * @param {import("./merge.js").InjectOptions} [options]
   * @returns {Promise<unknown|string>}
   */
  async injectJson(dataId, jsonValue, options = {}) {
    return this.injectJsonSync(dataId, jsonValue, options);
  }

  /**
   * @param {string} dataId
   * @param {unknown} jsonValue
   * @param {import("./merge.js").InjectOptions} [options]
   * @returns {unknown|string}
   */
  injectJsonSync(dataId, jsonValue, options = {}) {
    const base = this._requireStored(dataId);
    const merged = mergeJson(base, jsonValue, options.conflict);
    this._runTypeCheck(merged);
    this._store.set(dataId, merged);
    return formatInjectResult(merged, options);
  }

  /**
   * @param {string} dataId
   * @returns {unknown}
   */
  _requireStored(dataId) {
    if (typeof dataId !== "string" || dataId.length === 0) {
      throw new TypeError("dataId must be a non-empty string");
    }
    if (!this._store.has(dataId)) {
      throw new Error(`unknown data id: ${dataId}`);
    }
    return toMergeableJson(this._store.get(dataId));
  }

  /**
   * @param {string} dataId
   * @returns {Promise<unknown>}
   */
  async get(dataId) {
    return this.getSync(dataId);
  }

  /**
   * @param {string} dataId
   * @returns {unknown}
   */
  getSync(dataId) {
    if (typeof dataId !== "string" || dataId.length === 0) {
      throw new TypeError("dataId must be a non-empty string");
    }
    if (!this._store.has(dataId)) {
      throw new Error(`unknown data id: ${dataId}`);
    }
    const v = this._store.get(dataId);
    if (v instanceof XaiopFragment) {
      return new XaiopFragment(structuredClone(v.entries));
    }
    return structuredClone(v);
  }

  /**
   * @param {string} source
   * @param {boolean} [compatibilityMode=false] — omitted / false = strict;
   *   `true` = all fixes enabled
   * @returns {Promise<unknown>}
   */
  static async parse(source, compatibilityMode = false) {
    return parseAsync(source, compatibilityMode);
  }

  /**
   * @param {string} source
   * @param {boolean} [compatibilityMode=false] — omitted / false = strict;
   *   `true` = all fixes enabled
   * @returns {unknown}
   */
  static parseSync(source, compatibilityMode = false) {
    return parseSync(source, compatibilityMode);
  }

  /**
   * Encode a JSON value to XAIOP wire text (strict; no compat shapes).
   * @param {unknown} value
   * @param {import("./encode.js").EncodeOptions} [options]
   * @returns {Promise<string>}
   */
  static async encode(value, options = {}) {
    return encodeAsync(value, options);
  }

  /** Identical to {@link encode}; named to match free `encodeAsync`. */
  static async encodeAsync(value, options = {}) {
    return encodeAsync(value, options);
  }

  /**
   * @param {unknown} value
   * @param {import("./encode.js").EncodeOptions} [options]
   * @returns {string}
   */
  static encodeSync(value, options = {}) {
    return encodeValueSync(value, options);
  }

  /**
   * @param {unknown} baseJson
   * @param {string} xaiopSource
   * @param {import("./merge.js").MergeOptions} [options]
   */
  static mergeToJson(baseJson, xaiopSource, options = {}) {
    return mergeToJsonValue(baseJson, xaiopSource, options);
  }

  /**
   * @param {unknown} baseJson
   * @param {string} xaiopSource
   * @param {import("./merge.js").MergeToXaiopOptions} [options]
   */
  static mergeToXaiop(baseJson, xaiopSource, options = {}) {
    return mergeToXaiopValue(baseJson, xaiopSource, options);
  }

  /** @returns {boolean} */
  has(dataId) {
    return this._store.has(dataId);
  }

  /** @param {string} dataId */
  delete(dataId) {
    return this._store.delete(dataId);
  }

  clear() {
    this._store.clear();
  }
}

/** @param {number} seq @returns {string} */
function nextId(seq) {
  return `xaiop_${seq}_${Date.now().toString(36)}_${Math.random().toString(36).slice(2, 8)}`;
}

